src/compute/arithmetics/mod.rs

```diff
@@ -97,43 +91,66 @@ macro_rules! primitive_arithmetic_match {
-            "Arithmetic is currently only supported for arrays of the same logical type"
-                .to_string(),
-        ));
-    }
-    match data_type {
-        DataType::Int8 => primitive_arithmetic_match!(lhs, rhs, op, Int8Array),
-        DataType::Int16 => primitive_arithmetic_match!(lhs, rhs, op, Int16Array),
-        DataType::Int32 => primitive_arithmetic_match!(lhs, rhs, op, Int32Array),
-        DataType::Int64 | DataType::Duration(_) => {
-            primitive_arithmetic_match!(lhs, rhs, op, Int64Array)
+    use DataType::*;
+    use Operator::*;
+    match (lhs.data_type(), op, rhs.data_type()) {
+        (Int8, _, Int8) => primitive!(lhs, rhs, op, i8),
+        (Int16, _, Int16) => primitive!(lhs, rhs, op, i16),
+        (Int32, _, Int32) => primitive!(lhs, rhs, op, i32),
+        (Int64, _, Int64) | (Duration(_), _, Duration(_)) => {
+            primitive!(lhs, rhs, op, i64)
         }
-        DataType::UInt8 => primitive_arithmetic_match!(lhs, rhs, op, UInt8Array),
-        DataType::UInt16 => primitive_arithmetic_match!(lhs, rhs, op, UInt16Array),
-        DataType::UInt32 => primitive_arithmetic_match!(lhs, rhs, op, UInt32Array),
-        DataType::UInt64 => primitive_arithmetic_match!(lhs, rhs, op, UInt64Array),
-        DataType::Float16 => unreachable!(),
-        DataType::Float32 => primitive_arithmetic_match!(lhs, rhs, op, Float32Array),
-        DataType::Float64 => primitive_arithmetic_match!(lhs, rhs, op, Float64Array),
-        DataType::Decimal(_, _) => {
-            let lhs = lhs.as_any().downcast_ref::<Int128Array>().unwrap();
-            let rhs = rhs.as_any().downcast_ref::<Int128Array>().unwrap();
+        (UInt8, _, UInt8) => primitive!(lhs, rhs, op, u8),
+        (UInt16, _, UInt16) => primitive!(lhs, rhs, op, u16),
+        (UInt32, _, UInt32) => primitive!(lhs, rhs, op, u32),
+        (UInt64, _, UInt64) => primitive!(lhs, rhs, op, u64),
+        (Float32, _, Float32) => primitive!(lhs, rhs, op, f32),
+        (Float64, _, Float64) => primitive!(lhs, rhs, op, f64),
+        (Decimal(_, _), _, Decimal(_, _)) => {
+            let lhs = lhs.as_any().downcast_ref().unwrap();
+            let rhs = rhs.as_any().downcast_ref().unwrap();
 
             let res = match op {
-                Operator::Add => decimal::add::add(lhs, rhs),
-                Operator::Subtract => decimal::sub::sub(lhs, rhs),
-                Operator::Multiply => decimal::mul::mul(lhs, rhs),
-                Operator::Divide => decimal::div::div(lhs, rhs),
+                Add => decimal::add::add(lhs, rhs),
+                Subtract => decimal::sub::sub(lhs, rhs),
+                Multiply => decimal::mul::mul(lhs, rhs),
+                Divide => decimal::div::div(lhs, rhs),
             };
 
-            res.map(Box::new).map(|x| x as Box<dyn Array>)
+            res.map(|x| Box::new(x) as Box<dyn Array>)
+        }
+        (Time32(_), Add, Duration(_)) | (Date32, Add, Duration(_)) => {
```

Comment:
I think this is only valid for `TimeUnit::Millisecond` and `TimeUnit::Second`. The same goes for the subtraction

Reply:
great points. Addressed :)